### PR TITLE
Literals cleanups

### PR DESCRIPTION
**Pull request by xmo-odoo**
* cleans up unecessary `set()` calls, by replacing them with set literals or set comprehensions
* cleans up unnecessary `dict()` calls, by replacing them with dict literals or dict comprehensions, or the odd `dict.fromkeys`
* cleans up a few regex (unnecessary escapes) and implicit string concatenations (unnecessary continuations)

Originally partially in #73535  but that PR is too large and unfocused, so breaking it up into more specific cleanups (and also more complete thus this one being twice the size of that one),

### CODE REVIEW

**Comment by xmo-odoo** [comment]
One finding which arose from the dict cleanup is that there are two quite common patterns for which it might be useful to provide helpers:

1. Lots of `read_group(groupby=[a_field])` are immediately followed by collecting into a dict of `{a_field: a_field_count}`, or more rarely `{a_field: an_other_field}`, then dispatching this between records (that's generally when computing aggregates and such). Not sure how we'd call it but it seems like this could be some sort of `helper(field, value_field=field+'_count')` which would handle that internally.
2. A common missing piece is a `dict.fromkeys` which copies the value: lots of comprehensions are deployed in order to duplicate mutable values for all the keys e.g. `{rec: {} for rec in self}` and somesuch.

**Comment by Xavier-Do** [comment]
I'm not a big fan of this "cleanup", unnecessary is subjective

```python
dict(x='x', y='y', z='z')
```
 can be more readable than 
 ```python
 {'x': 'x', 'y': 'y', 'z': 'z'}
 ```
 depending on your editor because keys and values can be collored differently with the first example.
 
 I don't think imposing one style is a good idea.
 
image

**Comment by xmo-odoo** [comment]
> I'm not a big fan of this "cleanup", unnecessary is subjective
> 
> ```python
> dict(x='x', y='y', z='z')
> ```
> 
> can be more readable than
> 
> ```python
> {'x': 'x', 'y': 'y', 'z': 'z'}
> ```
> 
> depending on your editor because keys and values can be collored differently with the first example.

That is a very debatable point, and aside from that all the dict version does is increase cognitive complexity (more than one way to do things) at a strictly higher cost.

**Comment by Xavier-Do** [comment]
> That is a very debatable point

I agree, it is a debatable point, I just give an opinion ;)

>  at a strictly higher runtime cost.

If it is significant, it may be a good justification, but it could be explained in the pr description in this case ;)

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

Only touches sensitive sites to perform lexical updates.

**Review by AntoineVDV — COMMENTED** [review]
> Lots of read_group(groupby=[a_field]) are immediately followed by collecting into a dict of {a_field: a_field_count}, or more rarely {a_field: an_other_field}, then dispatching this between records (that's generally when computing aggregates and such). Not sure how we'd call it but it seems like this could be some sort of helper(field, value_field=field+'_count') which would handle that internally.

*That* would help clean up the code base of objectively unnecessary code. Indeed, most newbies' writing of their first `_compute_*_count` method goes like this:
1. be asked to batch the `search_count` for performance
2. look for a `search_count_group` for a good 10 min
3. realize it doesn't exist (yet? :crossed_fingers:) 
4. copy-paste from another `_compute_*_count` method and change the variable names

That would be cleaner to be able to do something like:
```py
model_b_count = self.env['model.b'].search_count_group('model_a_id', self.ids)
for model_a in self:
    model_a.model_b_count = model_b_count.get(model_a.id, 0)
```
instead of
```py
model_b_data = self.env['model.b'].read_group([('model_a_id', 'in', self.ids)], ['model_a_id'], ['model_a_id'])
model_b_count = {model_b_data['model_a_id'][0]: data['model_a_id_count'] for data in model_b_data}
for model_a in self:
    model_a.model_b_count = model_b_count.get(model_a.id, 0)
```

**Comment by tde-banana-odoo** [comment]
Hello,

I don't see the point of adding noise and breaking all current branches due to conflicts just for that ... what are you trying to fix ? Do we really gain something with that kind of changes ?

File: addons/payment_sips/tests/test_sips.py [hunk 1]
```diff
@@ -53,9 +53,9 @@ def test_redirect_form_values(self):
         return_url = self._build_url(SipsController._return_url)
         notify_url = self._build_url(SipsController._notify_url)
         self.assertEqual(form_inputs['Data'],
-            f'amount=111111|currencyCode=978|merchantId=dummy_mid|normalReturnUrl={return_url}|' \
-            f'automaticResponseUrl={notify_url}|transactionReference={self.reference}|' \
-            f'statementReference={self.reference}|keyVersion={self.sips.sips_key_version}|' \
+            f'amount=111111|currencyCode=978|merchantId=dummy_mid|normalReturnUrl={return_url}|'
+            f'automaticResponseUrl={notify_url}|transactionReference={self.reference}|'
+            f'statementReference={self.reference}|keyVersion={self.sips.sips_key_version}|'
             f'returnContext={json.dumps(dict(reference=self.reference))}'
```
[the remark this answers is not in the document]
Comment (xmo-odoo) [reply addons/payment_sips/tests/test_sips.py:59]: > ~I guess your grep brought you here for this particular line and you got distracted by the unnecessary `\` and forgot about the `dict()` call.~
> 
> Edit: just noticed the f-string.

No you're probably right that I missed it, apparently semgrep does not reach into f-strings when performing semantics analysis.

**Review by Julien00859 — COMMENTED** [review]
I'll have to agree with the other comments, while I am team literal, I don't really understand the value of changing that much code considering it'll be a pain forward-port wise. 

    ~ $ python3 -m timeit "{}"
    20000000 loops, best of 5: 18 nsec per loop
    ~ $ python3 -m timeit "dict()"
    5000000 loops, best of 5: 71.1 nsec per loop

Using literals we indeed gain *a little bit* but well... nanoseconds...

**Comment by rco-odoo** [comment]
I pretty much agree with the former comments.
While I am in favor of using list/set/dict comprehensions, this is not a valid reason to _impose_ a style.
And I don't see much added net value in this proposal, given the inconveniences it will cause when forward-porting fixes.

**Pull request closed without merging**